tests/testthat/test-crunch-data-frame.R

```diff
@@ -0,0 +1,184 @@
+        gndr <- as.vector(ds$gender)
+        expect_equal(nrow(ds_df), 25)
+        # both reording and subsetting the dataset
+        new_order <- c(4,3,1,2)
+        attr(ds_df, "order") <- new_order
+        expect_equal(ds_df$gender, gndr[new_order])
+        expect_equal(nrow(ds_df), 4)
+        ds_df2 <- as.data.frame(ds, row.order = new_order)
+        expect_equal(ds_df2$gender, gndr[new_order])
+        expect_equal(nrow(ds_df2), 4)
+    })
+    
+    test_that("Extract methods work with CrunchDataFrames like they do with data.frames", {
+        ds_df <- as.data.frame(ds)
+        true_df <- as.data.frame(ds, force = TRUE)
+        # single column/row extraction
+        expect_equivalent(ds_df[1,],true_df[1,])
+        expect_equivalent(ds_df[,1],true_df[,1])
+        
+        # multiple columns/rows
+        expect_equivalent(ds_df[c(1,2),],true_df[c(1,2),])
+        expect_equivalent(ds_df[,c(1,2)],true_df[,c(1,2)])
+        
+        # single column extraction with characters
+        expect_equivalent(ds_df[,"location"],true_df[,"location"])
+        
+        # multiple columns with characters
+        expect_equivalent(ds_df[,c("mymrset","textVar")],true_df[,c("mymrset","textVar")])
+        
+        # columns/rows with logicals
+        expect_equivalent(ds_df[,c(TRUE, FALSE, TRUE, FALSE, TRUE, FALSE, TRUE)],
+                          true_df[,c(TRUE, FALSE, TRUE, FALSE, TRUE, FALSE, TRUE)])
+        expect_equivalent(ds_df[c(TRUE, rep(FALSE, 23), TRUE),],
+                          true_df[c(TRUE, rep(FALSE, 23), TRUE),])
+        
+        # both rows and columns
+        expect_equivalent(ds_df[c(1,3,5),"location"],true_df[c(1,3,5),"location"])
+        
+        
+        # $ methods
+        expect_equivalent(ds_df$starttime,true_df$starttime)
+        
+        # [[ methods
+        expect_equivalent(ds_df[[1]],true_df[[1]])
+        expect_equivalent(ds_df[["catarray"]],true_df[["catarray"]])
+        expect_equivalent(ds_df[[2]],true_df[[2]])
+        expect_equivalent(ds_df[["birthyr"]],true_df[["birthyr"]])
+    })
+    
+    test_that("Extract methods input checking", {
+        ds_df <- as.data.frame(ds)
+        
+        expect_error(ds_df["foo",],
+                     "row subsetting must be done with either numeric or a logical")
+        expect_error(ds_df[,list(list(1))], 
+                     "column subsetting must be done with either numeric or a character")
+        
+        expect_error(ds_df[c(TRUE),], 
+                     paste0("when using a logical to subset rows, the logical vector ",
+                            "must have the same length as the number of rows"))
+        expect_error(ds_df[,c(TRUE)],
+                     paste0("when using a logical to subset columns, the logical ",
+                            "vector must have the same length as the number of columns"))
+        
+        expect_error(`$`(ds_df,1),
+                     paste0("invalid subscript type 'double'"))
```

Comment:
Is this just asserting how R handles bad input on `$`?

Reply:
Yeah, I wanted to make sure that we were covering that, and not getting something weird. Although we don't need to leave it in here since it really is just asserting that R does this for us, it might be nice to remind our future selves that we have checked that possibility.